File: research_server/app/snippet.py
"""Slice a long retrieved chunk down to one short, on-target snippet.

Gemini File Search keeps a whole sutta as ~one chunk, so the grounding chunk's
text is essentially the full sutta. For the human-facing "Sources" list we want a
short window around the matched terms — SQLite `snippet()`-style — not the whole
thing (the deep link opens the full text). This is pure: no SDK, no extra Gemini
call, so it adds no cost or latency.

Plan: docs/done/research/source-snippet-shortening-plan.md (§4, Option 1).
"""
from __future__ import annotations

import re

# Tiny English stopword set so scoring keys on content words ("eon", "mustard"),
# not glue ("the", "of", "is"). Deliberately small — the corpus is English
# (Sujato), and short words (<3 chars) are dropped anyway.
_STOPWORDS = frozenset(
    "and are but for from how its not that the their them then there these they "
    "this was were what when where which who why will with you your".split()
)

_WORD = re.compile(r"[A-Za-z]+")
# Sentence-ish split: end punctuation followed by whitespace. Pali prose is full
# of '.' '?' '"' so this is plenty to pick a window from.
_SENT_SPLIT = re.compile(r"(?<=[.?!])\s+")
_WS = re.compile(r"\s+")

# How far we'll nudge a cut to land on a word boundary instead of mid-word.
_SNAP = 24


def make_snippet(text: str, query: str, max_chars: int = 220) -> str:
    """Return a ~`max_chars` window of `text` around the `query` terms, with `…`.

    Falls back to the head of `text` when the query has no usable terms or none
    of them appear in the chunk. Returns "" only for empty input. The chunk's
    leading heading is split off upstream by `split_heading`, so `text` here is
    the sutta body — there's no reference prefix left to strip.
    """
    text = _WS.sub(" ", (text or "").strip())
    if not text:
        return ""
    if len(text) <= max_chars:
        return text

    terms = _query_terms(query)
    if not terms:
        return _head(text, max_chars)

    sentences = _SENT_SPLIT.split(text)
    best_i, best_score = 0, 0
    for i, sentence in enumerate(sentences):
        score = len(_words(sentence) & terms)
        if score > best_score:  # first sentence wins ties (>, not >=)
            best_i, best_score = i, score
    if best_score == 0:
        return _head(text, max_chars)

    # Grow a window of whole sentences around the best one, forward-biased.
    lo = hi = best_i
    length = len(sentences[best_i])
    while True:
        grew = False
        if hi + 1 < len(sentences) and length + 1 + len(sentences[hi + 1]) <= max_chars:
            hi += 1
            length += 1 + len(sentences[hi])
            grew = True
        if lo - 1 >= 0 and length + 1 + len(sentences[lo - 1]) <= max_chars:
            lo -= 1
            length += 1 + len(sentences[lo])
            grew = True
        if not grew:
            break

    window = " ".join(sentences[lo : hi + 1])
    left, right = lo > 0, hi < len(sentences) - 1
    # A single sentence can still exceed max_chars → trim around the first term.
    if len(window) > max_chars:
        window, tl, tr = _trim_around_terms(window, terms, max_chars)
        left, right = left or tl, right or tr
    return _wrap(window, left=left, right=right)


def split_heading(text: str, ref: str | None = None) -> tuple[str | None, str]:
    """Split a chunk at the ingest head/body boundary → (title, body).

    `ingest.load_unit` writes each unit as "<heading>\\n<body>", where the
    heading is the title segments (uid:0.*): collection name + number + chapter +
    sutta name, e.g. "Linked Discourses 15.6 Chapter One A Mustard Seed". We
    surface that as the citation's bold `title` (shown on its own line) and window
    only the body for the snippet — so every source shows a consistent heading
    instead of one that appears only when the match lands up top.

    `ref` (e.g. "SN 15.6") lets us drop just the number from the heading — it's
    already shown as the citation's `ref` — while keeping the collection name:
    "Linked Discourses Chapter One A Mustard Seed".

    Only the FIRST chunk of a sutta carries the ingest heading, which always leads
    with the collection ref number (e.g. "Linked Discourses 15.20 …"). Long suttas
    get split by File Search, so a grounding chunk can be a mid/tail slice that
    merely contains a newline; if the text before the first newline doesn't carry
    the ref number, it isn't a heading (it'd be e.g. a closing verse — see
    SN 15.20), so we keep the whole text as the body and emit no title. Returns
    (None, text) in that case and when there's no newline at all.
    """
    head, sep, body = text.partition("\n")
    num = _ref_number(ref)
    if not sep or not num or not re.search(rf"\b{re.escape(num)}\b", head):
        return None, text
    # Drop just the number (the citation already shows it as `ref`), keeping the
    # collection name: "Linked Discourses 15.6 Chapter One A Mustard Seed" ->
    # "Linked Discourses Chapter One A Mustard Seed". Group 1 = text before the
    # number; the number and its clinging space/punct are removed.
    title = re.sub(
        rf"^(.{{0,40}}?)\b{re.escape(num)}\b[\s.,:–—-]*", r"\1", head.strip(), count=1
    )
    return (_WS.sub(" ", title).strip() or None), body


def _ref_number(ref: str | None) -> str | None:
    """The numeric part of a display ref ("SN 15.20" -> "15.20"), else None."""
    m = re.search(r"\d+(?:[.\-]\d+)*", ref or "")
    return m.group() if m else None


def _query_terms(query: str) -> set[str]:
    return {
        w for w in _WORD.findall((query or "").lower())
        if len(w) > 2 and w not in _STOPWORDS
    }


def _words(text: str) -> set[str]:
    return set(_WORD.findall(text.lower()))


def _head(text: str, max_chars: int) -> str:
    """The opening of the chunk, word-boundary trimmed — the no-match fallback."""
    cut = text.rfind(" ", 0, max_chars)
    return _wrap(text[: cut if cut > 0 else max_chars], left=False, right=True)


def _trim_around_terms(s: str, terms: set[str], max_chars: int) -> tuple[str, bool, bool]:
    """Trim an over-long string to a `max_chars` window centred on the first term."""
    pos = next(
        (m.start() for m in _WORD.finditer(s.lower()) if m.group() in terms),
        0,
    )
    start = max(0, pos - max_chars // 2)
    end = min(len(s), start + max_chars)
    start = max(0, end - max_chars)  # re-pull start if we hit the right edge

    if start > 0:  # snap forward to a word start
        space = s.find(" ", start)
        if 0 <= space - start <= _SNAP:
            start = space + 1
    if end < len(s):  # snap back to a word end
        space = s.rfind(" ", start, end)
        if space != -1 and end - space <= _SNAP:
            end = space
    return s[start:end].strip(), start > 0, end < len(s)


def _wrap(window: str, *, left: bool, right: bool) -> str:
    window = window.strip()
    if left:
        window = "… " + window
    if right:
        window = window.rstrip(".,;:— ") + " …"
    return window

File: research_server/src/snippet.ts
// Snippet building is the CPU hotspot (Python worst case: 27ms). Budget levers
// used here: the body is truncated to a scan bound BEFORE any normalization,
// folded once per body with native string ops (no per-word regex walk), and
// terms are located with indexOf — a body with no term hit costs three native
// passes and nothing more. Callers dedupe by uid first, so all of this runs
// once per displayed source, not once per retrieved chunk.
//
// Kill switch if the 10ms budget is ever threatened: rung 1 — snippet only the
// uids that got a [[cite:]] chip (reorder buildResponse: finalize answer text
// first, then snippet the chipped set — typically 2–5 of ~14 sources; if zero
// chips, keep all so the cards aren't stripped bare). Rung 2 — skip makeSnippet
// entirely. `snippet: null` is contract-legal (prose-only cards already ship
// it) and rung-2 worst-case CPU is ~1–2ms.

const STOPWORDS = new Set(
  ('and are but for from how its not that the their them then there these they ' +
    'this was were what when where which who why will with you your').split(' '),
);

const WORD = /[\p{L}\p{M}]+/gu;
const MARKS = /\p{M}+/gu;
const WS = /\s+/g;
const LETTER = /\p{L}/u;

const SNAP = 24;
const ELLIPSIS_RESERVE = 4;
const MAX_SCAN_CHARS = 12_000;
const MAX_MATCHES = 24;
const MAX_MATCHES_PER_TERM = 8;
const FOLD_CACHE_MAX = 20_000;

// Fold = lowercase + strip diacritics ("Sāvatthī" → "savatthi"). The cache
// serves query terms and window words; whole bodies use foldText directly.
const foldCache = new Map<string, string>();

const foldText = (s: string): string =>
  s.toLowerCase().normalize('NFKD').replace(MARKS, '');

export function fold(word: string): string {
  let f = foldCache.get(word);
  if (f === undefined) {
    f = foldText(word);
    if (foldCache.size >= FOLD_CACHE_MAX) foldCache.clear();
    foldCache.set(word, f);
  }
  return f;
}

export function queryTerms(query: string): Set<string> {
  const terms = new Set<string>();
  for (const m of (query || '').matchAll(WORD)) {
    const folded = fold(m[0]);
    if (folded.length > 2 && !STOPWORDS.has(folded)) terms.add(folded);
  }
  return terms;
}

interface Match {
  start: number;
  end: number;
  term: string;
}

// indexOf per term over the folded body, word-boundary checked. Folded offsets
// drift from the original only where marks were stripped (sparse in this
// corpus); the window's word-snapping absorbs that.
function findTermMatches(folded: string, terms: Set<string>): Match[] {
  const out: Match[] = [];
  for (const term of terms) {
    let from = 0;
    let found = 0;
    while (found < MAX_MATCHES_PER_TERM) {
      const idx = folded.indexOf(term, from);
      if (idx === -1) break;
      const end = idx + term.length;
      const before = idx > 0 ? folded[idx - 1]! : '';
      const after = end < folded.length ? folded[end]! : '';
      if (!LETTER.test(before) && !LETTER.test(after)) {
        out.push({ start: idx, end, term });
        found++;
      }
      from = end;
    }
  }
  out.sort((a, b) => a.start - b.start);
  return out.length > MAX_MATCHES ? out.slice(0, MAX_MATCHES) : out;
}

export function makeSnippet(text: string, terms: Set<string>, maxChars = 220): string {
  // Bound the body first; whitespace-normalize only the emitted window (ingest
  // joins segments with single spaces, so a body-wide pass is wasted work).
  text = (text || '').slice(0, MAX_SCAN_CHARS).trim();
  if (!text) return '';
  if (text.length <= maxChars) return boldTerms(text.replace(WS, ' '), terms);

  const budget = Math.max(1, maxChars - ELLIPSIS_RESERVE);
  if (!terms.size) return head(text, budget);

  const matches = findTermMatches(foldText(text), terms);
  if (!matches.length) return head(text, budget);

  // Densest cluster: the match whose following budget-window covers the most
  // distinct terms wins; earliest wins ties.
  let bestI = 0;
  let bestScore = 0;
  for (let i = 0; i < matches.length; i++) {
    const limit = matches[i]!.start + budget;
    const seen = new Set<string>();
    for (let j = i; j < matches.length && matches[j]!.end <= limit; j++) {
      seen.add(matches[j]!.term);
    }
    if (seen.size > bestScore) {
      bestScore = seen.size;
      bestI = i;
    }
  }

  // Forward-biased window around the cluster, snapped to word boundaries.
  let start = Math.max(0, matches[bestI]!.start - Math.floor(budget / 3));
  let end = Math.min(text.length, start + budget);
  start = Math.max(0, end - budget);
  if (start > 0) {
    const sp = text.indexOf(' ', start);
    if (sp !== -1 && sp - start <= SNAP) start = sp + 1;
  }
  if (end < text.length) {
    const sp = text.lastIndexOf(' ', end);
    if (sp !== -1 && end - sp <= SNAP) end = sp;
  }
  const window = text.slice(start, end).replace(WS, ' ').trim();
  return boldTerms(wrap(window, start > 0, end < text.length), terms);
}

// Split a chunk at the ingest "<heading>\n<body>" boundary. Only a head chunk
// whose first line carries the ref number is a heading; the number itself is
// dropped from the title (the citation already shows it as `ref`).
export function splitHeading(
  text: string,
  ref?: string | null,
): { title: string | null; body: string } {
  const nl = text.indexOf('\n');
  const num = refNumber(ref);
  if (nl === -1 || !num) return { title: null, body: text };
  const headLine = text.slice(0, nl);
  const numRe = new RegExp(`\\b${escapeRe(num)}\\b`);
  if (!numRe.test(headLine)) return { title: null, body: text };
  const title = headLine
    .trim()
    .replace(new RegExp(`^(.{0,40}?)\\b${escapeRe(num)}\\b[\\s.,:–—-]*`), '$1')
    .replace(WS, ' ')
    .trim();
  return { title: title || null, body: text.slice(nl + 1) };
}

function refNumber(ref?: string | null): string | null {
  const m = /\d+(?:[.\-]\d+)*/.exec(ref ?? '');
  return m ? m[0] : null;
}

const escapeRe = (s: string): string => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

function boldTerms(window: string, terms: Set<string>): string {
  if (!terms.size) return window;
  const spans: [number, number][] = [];
  for (const m of window.matchAll(WORD)) {
    if (terms.has(fold(m[0]))) {
      spans.push([m.index, m.index + m[0].length]);
    }
  }
  for (let i = spans.length - 1; i >= 0; i--) {
    const [s, e] = spans[i]!;
    window = `${window.slice(0, s)}**${window.slice(s, e)}**${window.slice(e)}`;
  }
  return window;
}

function head(text: string, maxChars: number): string {
  const cut = text.lastIndexOf(' ', maxChars);
  return wrap(text.slice(0, cut > 0 ? cut : maxChars).replace(WS, ' '), false, true);
}

function wrap(window: string, left: boolean, right: boolean): string {
  window = window.trim();
  if (left) window = `… ${window}`;
  if (right) window = `${window.replace(/[.,;:— ]+$/, '')} …`;
  return window;
}
